File: Objects/TennisEvent.vb
<Serializable()> _
Public Class TennisEvent
    Inherits BusinessBase

    Private Shared log As ILog = LogManager.GetLogger(GetType(TennisEvent).ToString())

#Region "Events"
    'these events only occur when updating to the database
    'maybe they should happen when ApplyEdit is called?
    'maybe they should happen when IsDirtyChanged is raised?
    'hmmm....
    'maybe ApplyEdit is the best time, but how do we hook into that?

    Public Shared Event Created(ByVal obj As TennisEvent)
    Public Shared Event Updated(ByVal obj As TennisEvent)
    Public Shared Event Deleted(ByVal ID As Integer)

    Public Sub OnCreated()
        RaiseEvent Created(Me)
    End Sub

    Public Sub OnUpdated()
        RaiseEvent Updated(Me)
    End Sub

    Public Sub OnDeleted()
        RaiseEvent Deleted(Me.ID)
    End Sub

    Public Shared Sub OnDeleted(ByVal ID As Integer)
        RaiseEvent Deleted(ID)
    End Sub
#End Region

#Region "Private Data"
    Private _ID As Integer
    Private _Name As String = ""
    'Private _Classification As Classification
    Private _ClassificationID As Integer
    Private _BeginDate As New SmartDate(True)
    Private _EndDate As New SmartDate(False)
    Private _TeamPlay As Boolean
    Private _EventType As EventTypeEnum
#End Region

#Region "Business Properties and Methods"
    Public Enum EventTypeEnum As Short
        Normal = 0
        League = 1
        Tournament = 2
    End Enum

    Public ReadOnly Property ID() As Integer
        Get
            Return _ID
        End Get
    End Property

    Public Property Name() As String
        Get
            Return _Name
        End Get
        Set(ByVal value As String)
            If _Name = value Then Return
            _Name = value
            MarkDirty()
            ValidateProperty("Name", _Name)
            BrokenRules.Assert("EventNameRequired", "Please enter a name for this event", "Name", Name.Length = 0)
        End Set
    End Property

    Public Property ClassificationID() As Integer
        Get
            Return _ClassificationID
        End Get
        Set(ByVal value As Integer)
            If _ClassificationID = value Then Return
            _ClassificationID = value
            MarkDirty()
            ValidateProperty("ClassificationID", _ClassificationID)
            BrokenRules.Assert("SelectClassification", "Please select a classification", "ClassificationID", _ClassificationID = 0)
        End Set
    End Property

    Public Property BeginDate() As Date
        Get
            Return _BeginDate.Date
        End Get
        Set(ByVal value As Date)
            If _BeginDate.Date = value Then Return
            _BeginDate.Date = value
            MarkDirty()
            ValidateProperty("BeginDate", _BeginDate)
            'BrokenRules.Assert("SelectEvent", "Please select an event", "BeginDate", _EventID = 0)
        End Set
    End Property

    Public Property BeginDateText() As String
        Get
            Return _BeginDate.Text
        End Get
        Set(ByVal value As String)
            If _BeginDate.Text = value Then Return
            _BeginDate.Text = value
            MarkDirty()
            ValidateProperty("BeginDate", _BeginDate)
            'BrokenRules.Assert("SelectEvent", "Please select an event", "BeginDate", _EventID = 0)
        End Set
    End Property

    Public Property EndDate() As Date
        Get
            Return _EndDate.Date
        End Get
        Set(ByVal value As Date)
            If _EndDate.Date = value Then Return
            _EndDate.Date = value
            MarkDirty()
            ValidateProperty("EndDate", _EndDate)
            'BrokenRules.Assert("SelectEvent", "Please select an event", "EndDate", _EndDate = 0)
        End Set
    End Property

    Public Property EndDateText() As String
        Get
            Return _EndDate.Text
        End Get
        Set(ByVal value As String)
            If _EndDate.Text = value Then Return
            _EndDate.Text = value
            MarkDirty()
            ValidateProperty("EndDate", _EndDate)
            'BrokenRules.Assert("SelectEvent", "Please select an event", "EndDate", _EndDate = 0)
        End Set
    End Property

    Public Property TeamPlay() As Boolean
        Get
            Return _TeamPlay
        End Get
        Set(ByVal value As Boolean)
            If _TeamPlay = value Then Return
            _TeamPlay = value
            MarkDirty()
            'ValidateProperty("TeamPlay", _TeamPlay)
            'BrokenRules.Assert("SelectEvent", "Please select an event", "TeamPlay", _TeamPlay = 0)
        End Set
    End Property

    Public Property EventType() As EventTypeEnum
        Get
            Return _EventType
        End Get
        Set(ByVal value As EventTypeEnum)
            If _EventType = value Then Return
            _EventType = value
            MarkDirty()
            'ValidateProperty("EventType", _EventType)
        End Set
    End Property

    Public Property IsLeague() As Boolean
        Get
            Return EventType = EventTypeEnum.League
        End Get
        Set(ByVal value As Boolean)
            EventType = IIf(value, EventTypeEnum.League, EventTypeEnum.Normal)
        End Set
    End Property

    Public Property IsTournament() As Boolean
        Get
            Return EventType = EventTypeEnum.Tournament
        End Get
        Set(ByVal value As Boolean)
            EventType = IIf(value, EventTypeEnum.Tournament, EventTypeEnum.Normal)
        End Set
    End Property

    Public Overrides Function Save() As BusinessBase
        Dim creating As Boolean = IsNew
        Dim TennisEvent As TennisEvent = DirectCast(MyBase.Save(), TennisEvent)
        If creating Then
            TennisEvent.OnCreated()
        ElseIf IsNew Then
            TennisEvent.OnDeleted()
        Else
            TennisEvent.OnUpdated()
        End If
        Return TennisEvent
    End Function

#End Region

#Region "System.Object Overrides"
    Public Overrides Function ToString() As String
        'todo: return a useful string representation of the object
        Return MyBase.ToString()
    End Function

    Public Overloads Function Equals(ByVal A As String) As Boolean
        'todo: implement comparison logic
        Return False
    End Function

    Public Overrides Function GetHashCode() As Integer
        'todo: return a hash value for the object
        'for example, a common approach is to return ID.GetHashCode()
        Return MyBase.GetHashCode()
    End Function
#End Region

#Region "Shared Methods"
    Public Shared Function Create(Optional ByVal initialName As String = "") As TennisEvent
        Return New TennisEvent(initialName)
    End Function

    Public Shared Function Fetch(ByVal ID As Integer) As TennisEvent
        Return DirectCast(DataPortal.Fetch(New Criteria(ID)), TennisEvent)
    End Function

    Public Shared Function Fetch(ByVal dr As SafeDataReader) As TennisEvent
        Dim TennisEVent As New TennisEvent
        TennisEVent.MarkAsChild()
        TennisEVent._Fetch(dr)
        Return TennisEVent
    End Function

    Public Overloads Shared Sub Delete(ByVal ID As Integer)
        DataPortal.Delete(New Criteria(ID))
        OnDeleted(ID)
    End Sub
#End Region

#Region "Constructors"
    Private Sub New()
    End Sub

    Private Sub New(ByVal initialName As String)
        Name = initialName
    End Sub
#End Region

#Region "Criteria"
    <Serializable()> _
    Private Class Criteria
        Public ID As Int32
        Public Sub New(ByVal ID As Int32)
            Me.ID = ID
        End Sub
    End Class
#End Region

#Region "Data Access"
    Const spFetch As String = "csla_get_event"
    Const spUpdate As String = "csla_save_event"
    Const spDelete As String = "csla_delete_event"

    Private Sub _Fetch(ByVal dr As SafeDataReader)
        _ID = dr.GetInt32
        _Name = dr.GetString
        _ClassificationID = dr.GetInt32
        _BeginDate = dr.GetSmartDate(_BeginDate.EmptyIsMin)
        _EndDate = dr.GetSmartDate(_EndDate.EmptyIsMin)
        _EventType = DirectCast(dr.GetInt16, EventTypeEnum)
        _TeamPlay = dr.GetBoolean
    End Sub

    Protected Overrides Sub DataPortal_Fetch(ByVal Criteria As Object)
        Dim crit As Criteria = DirectCast(Criteria, Criteria)
		Dim dr As SafeDataReader = New SafeDataReader(DataAccess.ExecReader(spFetch, crit.ID))
        Try
            If Not dr.Read() Then
                Throw New ApplicationException(String.Format("Event not found (ID:{0})", crit.ID))
            End If
            _Fetch(dr)
            'If Not AreMaxLengthsSet Then
            '	LoadRules()
            'End If
            'log.Debug(String.Format("Business Rules: {0}{1}", ControlChars.CrLf, Rules))
        Finally
            dr.Close()
        End Try
        MarkOld()
    End Sub

    Protected Overrides Sub DataPortal_Update()
        'insert or update object data into database
        log.Debug("updating event " & ToString())

        Try
            If Me.IsDeleted Then
                If Not Me.IsNew Then
                    DataAccess.ExecUpdate(spDelete, ID)
                End If
                MarkNew()
            Else
                'DataAccess.GetParameter("@ID", IDParam(_ID, IsNew), ParameterDirection.InputOutput),
                Dim ret As ArrayList = DataAccess.ExecUpdate(spUpdate, _
                      IDParam(_ID, IsNew), _
                     _Name, _
                     _ClassificationID, _
                     _BeginDate.DBValue, _
                     _EndDate.DBValue, _
                     _EventType, _
                     _TeamPlay)
                If IsNew Then
                    _ID = CInt(ret(0))
                    log.Info(String.Format("New id is {0}", _ID))
                End If
                MarkOld()
            End If
        Catch ex As Exception
            log.Error("error in DataPortal_Update", ex)
            Throw New Library.DataPortalException("error in DataPortal_Update", ex)
        End Try

    End Sub

    Protected Overrides Sub DataPortal_Delete(ByVal Criteria As Object)
        Dim crit As Criteria = DirectCast(Criteria, Criteria)
        Try
            DataAccess.ExecUpdate(spDelete, crit.ID)
        Catch ex As Exception
            log.Error("error in DataPortal_Delete", ex)
            Throw New Library.DataPortalException("error in DataPortal_Delete", ex)
        End Try
    End Sub
#End Region

End Class

File: TennisObjects/TennisEvent.cs
using System;
using System.Collections.Generic;
using System.Text;
using Csla;
using log4net;
using Bahns.Data;
using Bahns.Util;
using System.Collections;

namespace TennisObjects
{
	public enum EventTypeEnum : short
	{
		All = -1, //used for querying
		Normal = 0,
		League = 1,
		Tournament = 2
	}

	[Serializable()]
	public class TennisEvent : BusinessBase<TennisEvent>
	{

		private static ILog log = LogManager.GetLogger(typeof(TennisEvent).ToString());

		#region Events
		//these events only occur when updating to the database 
		//maybe they should happen when ApplyEdit is called? 
		//maybe they should happen when IsDirtyChanged is raised? 
		//hmmm.... 
		//maybe ApplyEdit is the best time, but how do we hook into that? 

		public static event CreatedEventHandler Created;
		public delegate void CreatedEventHandler(TennisEvent obj);
		public static event UpdatedEventHandler Updated;
		public delegate void UpdatedEventHandler(TennisEvent obj);
		public static event DeletedEventHandler Deleted;
		public delegate void DeletedEventHandler(int ID);

		public void OnCreated()
		{
			if (Created != null)
			{
				Created(this);
			}
		}

		public void OnUpdated()
		{
			if (Updated != null)
			{
				Updated(this);
			}
		}

		public void OnDeleted()
		{
			if (Deleted != null)
			{
				Deleted(this.ID);
			}
		}

		public static void OnDeleted(int ID)
		{
			if (Deleted != null)
			{
				Deleted(ID);
			}
		}
		#endregion

		#region Fields
		private int _ID;
		private string _Name = "";
		//Private _Classification As Classification 
		private int _ClassificationID;
		private SmartDate _BeginDate; // = new SmartDate(true);
		private SmartDate _EndDate; // = new SmartDate(false);
		private bool _TeamPlay;
		private EventTypeEnum _EventType;
		#endregion

		#region Properties
		public int ID
		{
			get { return _ID; }
		}

		public string Name
		{
			get { return _Name; }
			set { SetProperty(ref _Name, value); }
		}

		public int ClassificationID
		{
			get { return _ClassificationID; }
			set { SetProperty(ref _ClassificationID, value); }
		}

		public SmartDate BeginDate
		{
			get { return _BeginDate; }
			set { SetProperty(ref _BeginDate, value); }
		}

		public string BeginDateText
		{
			get { return _BeginDate.Text; }
			set { BeginDate = SmartDate.FromString(value);}
		}

		public SmartDate EndDate
		{
			get { return _EndDate; }
			set { SetProperty(ref _EndDate, value); }
		}

		public string EndDateText
		{
			get { return _EndDate.Text; }
			set { EndDate = SmartDate.FromString(value); }
		}

		public bool TeamPlay
		{
			get { return _TeamPlay; }
			set { SetProperty(ref _TeamPlay, value); }
		}

		public EventTypeEnum EventType
		{
			get { return _EventType; }
			set { SetProperty(ref _EventType, value); }
		}

		public bool IsLeague
		{
			get { return EventType == EventTypeEnum.League; }
			set { EventType = (value ? EventTypeEnum.League : EventType); }
		}

		public bool IsTournament
		{
			get { return EventType == EventTypeEnum.Tournament; }
			set { EventType = (value ? EventTypeEnum.Tournament : EventType); }
		}

		public bool IsCurrent
		{
			get 
			{
				return (BeginDate.IsNull || BeginDate < DateTime.Now) && (EndDate.IsNull || EndDate > DateTime.Now);
			}
		}
	

		public override TennisEvent Save()
		{
			bool creating = IsNew;
			TennisEvent tennisEvent = base.Save();
			if (creating)
			{
				tennisEvent.OnCreated();
			}
			else if (IsNew)
			{
				tennisEvent.OnDeleted();
			}
			else
			{
				tennisEvent.OnUpdated();
			}
			return tennisEvent;
		}

		#endregion

		#region System.Object Overrides
		protected override object GetIdValue()
		{
			return _ID;
		}
		#endregion

		#region Static Methods
		public static TennisEvent Create() // ERROR: Unsupported modifier : In, Optional string initialName) 
		{
			return Create("");
		}

		public static TennisEvent Create(string initialName) // ERROR: Unsupported modifier : In, Optional string initialName) 
		{
			return new TennisEvent(initialName);
		}

		public static TennisEvent Fetch(int ID)
		{
			return (TennisEvent)DataPortal.Fetch(new Criteria(ID));
		}

		internal static TennisEvent Fetch(SafeDataReader dr)
		{
			TennisEvent TennisEVent = new TennisEvent();
			TennisEVent.MarkAsChild();
			TennisEVent._Fetch(dr);
			return TennisEVent;
		}

		public static void Delete(int ID)
		{
			DataPortal.Delete(new Criteria(ID));
			OnDeleted(ID);
		}
		#endregion

		#region Constructors
		private TennisEvent()
		{
		}

		private TennisEvent(string initialName)
		{
			Name = initialName;
		}
		#endregion

		#region Criteria
		[Serializable()]
		private class Criteria
		{
			public Int32 ID;
			public Criteria(Int32 ID)
			{
				this.ID = ID;
			}
		}
		#endregion

		#region Data Access
		private void _Fetch(SafeDataReader dr)
		{
			_ID = dr.GetInt32();
			_Name = dr.GetString();
			_ClassificationID = dr.GetInt32();
			_BeginDate = dr.GetDateTime();
			_EndDate = dr.GetDateTime();
			_EventType = (EventTypeEnum)dr.GetInt16();
			_TeamPlay = dr.GetBoolean();
		}

		protected override void DataPortal_Fetch(object Criteria)
		{
			Criteria crit = (Criteria)Criteria;
			using (SafeDataReader dr = DbHelper.ExecuteReader("csla_get_event", crit.ID))
			{
				if (!dr.Read())
				{
					throw new RecordNotFoundException(string.Format("Event not found (ID:{0})", crit.ID));
				}
				_Fetch(dr);
			}
			MarkOld();
		}

		protected override void DataPortal_Insert()
		{
			Util.CheckAddAccess();

			//insert or update object data into database 
			log.Debug("updating event " + ToString());

			DbResultsDictionary ret = DbHelper.ExecuteNonQuery("csla_save_event", Util.IDParam(_ID, IsNew), _Name, _ClassificationID, _BeginDate.DbValue, _EndDate.DbValue, _EventType, _TeamPlay);
			_ID = (int)ret["@ID"];
			log.Info(string.Format("New id is {0}", _ID));
		}

		protected override void DataPortal_Update()
		{
			Util.CheckEditAccess();

			//insert or update object data into database 
			log.Debug("updating event " + ToString());

				DbResultsDictionary ret = DbHelper.ExecuteNonQuery("csla_save_event", Util.IDParam(_ID, IsNew), _Name, _ClassificationID, _BeginDate.DbValue, _EndDate.DbValue, _EventType, _TeamPlay);
				if (IsNew)
				{
					_ID = (int)ret["@ID"];
					log.Info(string.Format("New id is {0}", _ID));
				}
		}

		protected override void DataPortal_DeleteSelf()
		{
			Util.CheckDeleteAccess();
			DbHelper.ExecuteNonQuery("csla_delete_event", ID);
		}

		protected override void DataPortal_Delete(object Criteria)
		{
			Criteria crit = (Criteria)Criteria;
			DbHelper.ExecuteNonQuery("csla_delete_event", crit.ID);
		}
		#endregion

	}

}
